### PR TITLE
FP16 Support - Change Set 2

### PR DESCRIPTION
**Pull request by deven-amd**
Changes include

1. enable all "half2" related code on the GPU.
2. enable "ScanLauncher" on the GPU.

### CODE REVIEW

File: third_party/eigen3/eigen.patch [hunk 1]
```diff
@@ -5702,7 +5577,7 @@ diff -Naur eigen/eigen-eigen-429aa5254200/unsupported/Eigen/CXX11/src/Tensor/Ten
 +  }
 +};
 +
-+#ifdef EIGEN_HAS_HIP_FP16_HALF2
++#ifdef EIGEN_HAS_HIP_FP16_HALF
```
[the remark this answers is not in the document]
Comment (deven-amd) [reply third_party/eigen3/eigen.patch]: Yes.

The EIGEN_HAS_HIP_FP16_HALF2 macro was never defined.
It was only intended to comment out code that 
1. had references to "half2"
AND
2. I had not had the opportunity to update for HIP fp16.

This commit updates the "half2" references to correctly use the HIP fp16 version (where needed)

File: third_party/eigen3/eigen.patch [hunk 2]
```diff
@@ -1761,42 +1672,6 @@ diff -Naur eigen/eigen-eigen-429aa5254200/Eigen/src/Core/arch/HIP/hcc/PacketMath
...
-+
-+template<> EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE half2 plog<half2>(const half2& a) {
-+  float a1 = __hip_low2float(a);
-+  float a2 = __hip_high2float(a);
-+  float r1 = logf(a1);
-+  float r2 = logf(a2);
-+  return __hip_floats2half2_rn(r1, r2);
-+}
-+
-+template<> EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE half2 pexp<half2>(const half2& a) {
```
[the remark this answers is not in the document]
Comment (deven-amd) [reply third_party/eigen3/eigen.patch]: They are not really omitted, they just do not require any changes :)

Those functions take "half2" as input and return "half2". hip_fp16.h has native implementation for them, and so we never need to do the "half2 <---> Eigen::half" conversions and hence no change

[the remark this answers is not in the document]
Comment (whchung) [reply third_party/eigen3/eigen.patch]: Thanks. Now it makes sense to me :+1:

File: third_party/eigen3/eigen.patch (hunk 1 above)
[the remark this answers is not in the document]
Comment (whchung) [reply third_party/eigen3/eigen.patch]: got it

File: tensorflow/core/framework/tensor.cc [hunk 3]
```diff
@@ -360,7 +360,11 @@ struct ProtoHelper<Eigen::half> {
   static void Fill(const Eigen::half* data, size_t n, TensorProto* proto) {
     proto->mutable_half_val()->Reserve(n);
     for (size_t i = 0; i < n; ++i) {
+#if defined(TENSORFLOW_USE_ROCM_HIP_FP16)      
+      proto->mutable_half_val()->AddAlreadyReserved(__half_as_ushort(data[i].x));
+#else
       proto->mutable_half_val()->AddAlreadyReserved(data[i].x);
+#endif
```
Comment (whchung) [inline tensorflow/core/framework/tensor.cc]: @deven-amd For future reference, logic here means:

- for kernel path (upper one, `#ifdefined(TENSORFLOW_USE_ROCM_HIP_FP16)` ), we cast from native fp16 type to unsigned short on host side
- for host path (bottom one, `#else`), we simply reuse `.x` which is already an unsigned short

so no matter what path we take on host side fp16 are always treated as unsigned short, is this understanding correct?

File: third_party/eigen3/eigen.patch [hunk 4]
```diff
@@ -6234,3 +6118,63 @@ diff -Naur eigen/eigen-eigen-429aa5254200/unsupported/Eigen/CXX11/src/Tensor/Ten
...
+diff -Naur eigen/eigen-eigen-429aa5254200/unsupported/Eigen/CXX11/src/Tensor/TensorIndexList.h eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorIndexList.h
+--- eigen/eigen-eigen-429aa5254200/unsupported/Eigen/CXX11/src/Tensor/TensorIndexList.h	2017-09-20 04:22:23.000000000 -0400
++++ eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorIndexList.h	2018-05-02 12:40:34.829676980 -0400
+@@ -350,7 +350,8 @@
+ 
+ namespace internal {
+ 
+-template<typename FirstType, typename... OtherTypes> size_t array_prod(const IndexList<FirstType, OtherTypes...>& sizes) {
++template<typename FirstType, typename... OtherTypes>
++  EIGEN_STRONG_INLINE EIGEN_DEVICE_FUNC size_t array_prod(const IndexList<FirstType, OtherTypes...>& sizes) {
```
Comment (whchung) [inline third_party/eigen3/eigen.patch]: Thanks for fixing this. I believe @sunway513 were haunted by this awhile ago :)

File: tensorflow/core/framework/tensor.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/core/framework/tensor.cc]: Correct.

File: third_party/eigen3/eigen.patch (hunk 4 above)
Comment (deven-amd) [reply third_party/eigen3/eigen.patch]: u r welcome :)

**Pull request merged**